### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by reisub-de**
- Auto reformat
- some code cleanup 
- some really minor fixes
- first tests added

### CODE REVIEW

File: dmpr.py [hunk 1]
```diff
@@ -63,28 +65,28 @@ def log(self, tracepoint, msg):
 
 
 class DMPR(object):
-
+    # TODO: change signature, log is required
     def __init__(self, log=None, tracer=None):
-        assert(log)
+        assert (log)
```
Comment (hgn) [inline dmpr.py:70]: sure? Whitespace before parentheses?

File: dmpr.py [hunk 2]
```diff
@@ -248,45 +240,40 @@ def tick(self):
             self._recalculate_routing_table()
 
         now = self._get_time(priv_data=self._get_time_priv_data)
-        self.tracer.log(self.tracer.TICK, {'now' : now})
+        self.tracer.log(self.tracer.TICK, {'now': now})
```
Comment (hgn) [inline dmpr.py:243]: without whitspace, is this defined in pep8? not sure, just asking ....

File: dmpr.py [hunk 3]
```diff
@@ -248,45 +240,40 @@ def tick(self):
...
             self.transmitted_now = True
         else:
             self.transmitted_now = False
 
+    def stop(self):
+        self._reset()
+        now = self._get_time(priv_data=self._get_time_priv_data)
+        self.log.warning("stop DMPR core", time=now)
 
-    def stop(self, init=False):
```
Comment (hgn) [inline dmpr.py:260]: mhh, is this tested and checked in dmpr*d*?

File: dmpr.py [hunk 4]
```diff
@@ -248,45 +240,40 @@ def tick(self):
...
-
     def start(self):
         now = self._get_time(priv_data=self._get_time_priv_data)
         self.log.info("start DMPR core", time=now)
-        assert(self._get_time)
-        assert(self._routing_table_update_func)
-        assert(self._packet_tx_func)
-        assert(self._conf)
-        assert(self._routing_table == None)
+        assert self._get_time
```
Comment (hgn) [inline dmpr.py:264]: here without parentheses ....

File: dmpr.py [hunk 5]
```diff
@@ -323,19 +307,16 @@ def _calc_next_tx_time(self):
...
 
-
     def _is_valid_interface(self, interface_name):
-        found = False
         for interface in self._conf["interfaces"]:
             if interface_name == interface["name"]:
-                found = True
-        return found
-
+                return True
```
Comment (hgn) [inline dmpr.py:313]: nice catch, thanks

File: dmpr.py [hunk 6]
```diff
@@ -456,224 +432,213 @@ def _recalculate_routing_table(self):
...
-                         break
-                  if path_found == False:
-                     neigh_data['paths']["{}>{}".
-                                         format(self._conf["id"],neigh_name)].append(interface_name)
-                  found_neigh = True
-                  break
-           if found_neigh == False:
-              neigh_routing_paths = self._add_neigh_entries(interface_name, sender_name, 
-                                                            sender_data, neigh_routing_paths)
+            for neigh_name, neigh_data in neigh_routing_paths['neighs'].items():
```
Comment (hgn) [inline dmpr.py:470]: great improvement, but I think we can improve it even further. I.e. split in separate functions, etc ..

File: dmpr.py [hunk 7]
```diff
@@ -456,224 +432,213 @@ def _recalculate_routing_table(self):
...
-           othernode_bw[sender_name]['path_characteristics'] = dict()
+            othernode_bw[sender_name] = dict()
+            othernode_bw[sender_name]['path_characteristics'] = dict()
         else:
-             self.log.info('updating high bandwidth other nodes from the existing neighbour')
+            self.log.info('updating high bandwidth other nodes from the existing neighbour')
         othernode_bw[sender_name] = sender_data['msg']['routingpaths']['high_bandwidth'].copy()
-        othernode_bw[sender_name]['path_characteristics'] = sender_data['msg']['routingpaths']['path_characteristics'].copy()
+        othernode_bw[sender_name]['path_characteristics'] = sender_data['msg']['routingpaths'][
+            'path_characteristics'].copy()
```
Comment (hgn) [inline dmpr.py:510]: not sure if copy is really required or if the whole datastructure can be better coded ... Please check

File: dmpr.py [hunk 8]
```diff
@@ -456,224 +432,213 @@ def _recalculate_routing_table(self):
...
             for neigh_path_name in neigh_paths_name:
                 for interface_name in self._conf['interfaces']:
                     if interface_name['name'] == neigh_path_name:
-                       if len(loss_dict)>0:
-                          for path_name, path_weight in loss_dict.items():
-                              if interface_name['link-characteristics']['loss'] < path_weight:
-                                 loss_dict = dict()
-                                 loss_dict[interface_name['name']] = interface_name['link-characteristics']['loss']
-                       else:
+                        if len(loss_dict) > 0:
```
Comment (hgn) [inline dmpr.py:564]: good improvement, but as said earlier, the whole block locks terrible ... ;-)

File: dmpr.py [hunk 9]
```diff
@@ -456,224 +432,213 @@ def _recalculate_routing_table(self):
...
         return compressedloss
 
-
     def add_bandwidth_entry(self, neigh_name, neigh_data, weigh_bandwidth, compressedBW):
-        route="{}>{}".format(self._conf["id"], neigh_name)
-        compressedBW[neigh_name] = {'next-hop':neigh_data['next-hop'],
-                                    'networks':neigh_data['networks']
-                                   }
+        route = "{}>{}".format(self._conf["id"], neigh_name)
+        compressedBW[neigh_name] = {'next-hop': neigh_data['next-hop'],
```
Comment (hgn) [inline dmpr.py:604]: camelCase should be replace by snake_case sometimes ...

File: dmpr.py [hunk 10]
```diff
@@ -724,195 +686,191 @@ def _map_BW_values(self, node_name, weight_update, dest_name, dest_data):
...
-                                  if path_char['loss']==fib_path_data['loss'] and path_char['bandwidth']==fib_path_data['bandwidth']:
-                                     path_found = True
-                                     loss_entry['interface'] = interface['name']
-                                     break
-                              if path_found==True:
-                                 break
-                       break
-                loss_entry['next-hop'] = self.next_hop_ip_addr(loss_entry['proto'], dest_data['next-hop'], loss_entry['interface'])
+                    if path_dir == search_key:
+                        for fib_path_name, fib_path_data in self.fib['path_characteristics'].items():
```
Comment (hgn) [inline dmpr.py:815]: I think this is the place where the block should be separated into a own method. But wait until you refactored the whole code and we understand where we wanted to go ...

**Review by hgn — COMMENTED** [review]
Great changeset! Just added some notes ... I approve this MR but please check the comments.

Thank you!

File: dmpr.py (hunk 1 above)
Comment (reisub-de) [reply dmpr.py:70]: parentheses is not even needed, this was just from the first round of auto pep8, will be obsolete soon anyway

File: dmpr.py (hunk 2 above)
Comment (reisub-de) [reply dmpr.py:243]: yep

File: dmpr.py (hunk 3 above)
Comment (reisub-de) [reply dmpr.py:260]: the init parameter was only used internally

File: dmpr.py (hunk 6 above)
Comment (reisub-de) [reply dmpr.py:470]: see next pr coming soon

File: dmpr.py (hunk 7 above)
Comment (reisub-de) [reply dmpr.py:510]: will be obsolete soon

**Pull request merged**